### DISCUSSION TITLE
Setting up Mirth in Eclipse

### DISCUSSION THREAD

**Discussion by BluetonicUK** [nextgenhealthcare/connect#5156] [Q&A]
Hi, 

Apologies if this has been answered. I've trawled the internet looking for tutorials but they are outdated. I come from a .NET background so I'm used to working with VS using C#. I'm struggling getting Mirth to build in Eclipse so we can build our own extensions etc. 

I'm unable to execute the ant-build due to the following errors: 

![image](https://user-images.githubusercontent.com/59801661/165312771-996bd4b3-d5e3-4d9e-b7c2-816793e2f075.png)

I just wondered if there is a link somewhere that's up to date to get things working in the IDE.
Thanks

**Comment by ab-20-m** [comment]
https://github.com/nextgenhealthcare/connect/wiki/Clone-Build-and-Run-from-Source#building-mirth-connect

**Reply by siddharthsharma9258** [reply]
Great stuff!

**Comment by jonbartels** [comment]
> I'm struggling getting Mirth to build in Eclipse **so we can build our own extensions** etc.

You do not need to build all of MC to write extensions. Extensions can exist in their own projects and include MC JARs as dependencies.

https://github.com/kpalang/mirth-plugin-guide

**Reply by siddharthsharma9258** [reply]
Hey, I have a question.

Any one can contribute towards making extensions and plugins? Is it like eclipse extensions, kind of download on a need to have basis?

**Reply by pacmano1** [reply]
Yup.  You can even sell them it you like.

**Reply by jonbartels** [reply]
Yes you can. Check out these authors:

- https://github.com/kpalang/ is the leader in open source MC plugin development. His guides are good. 
- https://github.com/kayyagari/ does excellent development. He has not documented his build process as well as kpalang but his work is excellent
- https://github.com/tonygermano/connect-plugins also good development. He uses Kotlin which runs in the JVM 
- https://consultzen.com/zen-ssl-extension/ publishes an SSL extension for MC for sale. 

Finally MC itself uses extensions for the core functionality anyway. The File Reader connector is an extension, the HTTP sender is an extension, etc.

**Reply by siddharthsharma9258** [reply]
Is there a validation process for extensions made like that, something like it is safe to use?

**Reply by pacmano1** [reply]
that's a fine idea.

**Reply by jonbartels** [reply]
@siddharthsharma9258 
> Is there a validation process for extensions made like that, something like it is safe to use?

Yes. The plugins are supposed to be signed. This means that they are compiled by a known person/business whose identity was verified by the certificate issuer. See https://github.com/kpalang/mirth-plugin-guide#---signing-and-publishing for more information.

**Reply by pacmano1** [reply]
Known business != vetted by anyone as to suitability to function, security,  malicious function, etc. of course.

**Comment by BluetonicUK** [comment]
Still having issues with this the same as my original screenshot. I have tried different JDK versions, reinstalled Eclipse many times. The same errors persist:


Project 'Client' is missing required library: '/Server/setup/client-lib/userutil-sources.jar'	Client		Build path	Build Path Problem

and


Project 'Server' is missing required source folder: 'build'	Server		Build path	Build Path Problem


Am I doing something really stupid? Because this doesn't make sense to me at all. The cloned repo should just build using ant in Eclipse to my knowledge

**Reply by ab-20-m** [reply]
Is Ant failing to execute Server/mirth-build.xml? That build script should generate the missing artifacts that Eclipse complains about in the build paths.

**Reply by BluetonicUK** [reply]
> Is Ant failing to execute Server/mirth-build.xml? That build script should generate the missing artifacts that Eclipse complains about in the build paths.

I set the JRE to point to the 1.8 folder as described in the guide and just double checked the compiler window to ensure that was the same. 

I get this in the build log:

BUILD FAILED
C:\Users\blabla\Desktop\MirthRepo\connect\server\mirth-build.xml:37: The following error occurred while executing this line:
C:\Users\blabla\Desktop\MirthRepo\connect\server\build.xml:82: Compile failed; see the compiler error output for details.

The above are clickable and produce these snippets respectively:

![image](https://user-images.githubusercontent.com/59801661/168616799-d51ce41a-f93d-42f2-9c39-534b70ff9e36.png)

and

![image](https://user-images.githubusercontent.com/59801661/168616951-42afc99d-9c84-4cbd-b479-bd8b6f5a5a62.png)

		

Not sure if that helps or not?

**Reply by ab-20-m** [reply]
Did you configure Ant to use a JRE or a JDK? Java 8 JREs don't have javac included so the failure is expected.

**Reply by BluetonicUK** [reply]
> Did you configure Ant to use a JRE or a JDK? Java 8 JREs don't have javac included so the failure is expected.

I have it configured to use JDK 1.8. I have tried different versions though. 

I realised last night that running `javac -version` and `java -version` in CMD that the versions were different, I managed to get them on the same version with lots of messing around but I'd made lots of changes by this point. I think you've helped me identify the issue though. I need to investigate further. 

May I ask what version of java & javac you have success with?

**Reply by BluetonicUK** [reply]
I've been doing some digging. Changed my Java version to the latest one and pointing at the JDK in the ANT Build. 

First step was to uncomment all the Java 9+ info and add the dashes to the arguments. 

The error I now get is: [javac] error: module not found: java.xml.ws in this line (there are 2 more lines like this):
![image](https://user-images.githubusercontent.com/59801661/168828103-233af179-589d-497c-892d-b9e69d6aaa5d.png)

I tried to remove the ',java.xml.ws' from all lines and it does build up to the Junit test and fails there. I will keep investigating. I can'[t understand why I am getting this issues and others just seem to load it fine.

EDIT: java.xml.ws is deprecated and has been replaced by https://docs.oracle.com/javase/10/docs/api/java.xml.ws-summary.html - this needs changing in the repo.

**Reply by ab-20-m** [reply]
It looks like the Build instructions might not have been recently tested against newer Eclipse releases that require JDK 11+.

Here's my "fresh" setup instructions that hopefully work for you, at least through the build/package parts JUnit tests are bombing out during the JVM process fork and I'm too lazy to modify the Ant script/figure out why at the moment. Versions are just what I used, later releases "should" work.

1. Install Eclipse 2022-03, select the default bundled JRE.
2. Install Amazon Corretto JDK 8u332, do not enable the "Environment Setup" feature.
3. Select the Import Git Project, use the clone URI, import existing Eclipse Projects
4. Select Window -> Preferences -> Java -> Installed JREs; Create option for the JDK8 install, select it and deselect the default "jre", Select Apply.
5. Select Window -> Preferences -> Java -> Compiler, Set Compliance level to 1.8, Select Apply.
6. Right click on Server/mirth-build.xml -> Ant build... -> Main -> Arguments, Enter: -Dant.build.javac.source=1.8 -Dant.build.javac.target=1.8 , Select Apply.
7. Select JRE -> "Run in the same JRE as the Workspace", Apply
8. Select Run

Another alternative is do the ant build completely outside of Eclipse which I think is what I did the last time I had to setup an environment and got tired of fighting with Eclipse.

**Reply by ab-20-m** [reply]
Actually this looks like something changed/broke in Eclipse between Eclipse 2021-12 and 2022-03. Eclipse 2021-12 I can make a simple Ant build.xml that runs against a different JRE/JDK than Eclipse uses, works fine when I launch. When I switch over to Eclipse 2022-03 I start getting JNI Errors, specifically an SocketTimeoutException: Accept timed out. The default for Ant JRE timeouts is 20 seconds so this shouldn't be happening as fast as it is.

Edit:
See [Eclipse Bug 579317](https://bugs.eclipse.org/bugs/show_bug.cgi?id=579317). In Eclipse 2022-03 the Ant plugin was upgraded to compile against Java 11 and that breaks trying to run Ant against Java 8. Anyone using latest Eclipse will have to install and configure Ant outside of Eclipse and run the scripts at the command line.

**Reply by BluetonicUK** [reply]
Thank you for all your guidance & help with this. I've been manic with other work today but will try this later this week :) 

Thank you again.

**Reply by cturczynskyj** [reply]
I ran into this today. I was able to get run an ANT build in Eclipse by using the External Tools settings and configuring a new Program.
![Screen Shot 2022-06-01 at 4 14 16 PM](https://user-images.githubusercontent.com/15792647/171516657-d3fe809c-5831-4e98-8ec0-4ccf83dc03d6.png)
![Screen Shot 2022-06-01 at 4 15 29 PM](https://user-images.githubusercontent.com/15792647/171516717-17e441c3-d186-4ddf-b949-740fc3cc93c3.png)
